File: backend/routes/stripe.py
import os
from flask import Blueprint, jsonify, request
from services.auth import token_required
from services.stripe_client import (
    publishableKey,
    create_booking_payment,
    release_booking_hold,
    finalize_booking,
    createConnectAccount,
    createAccountLink,
    sync_connect_account,
    handle_webhook,
    onboardingComplete,
)
from services.payouts import run_payout_sweep
from services.notifications import run_notification_sweep

stripe_bp = Blueprint('stripe', __name__)


@stripe_bp.route('/stripe/key', methods=['GET'])
def getKey():
    return jsonify({"publishableKey": publishableKey}), 200


# Atomically reserve the slot and start a platform-held payment for a booking.
# Funds are held on SpotOn's balance and transferred to the seller later.
@stripe_bp.route('/stripe/create-booking-payment', methods=['POST'])
@token_required
def create_booking_payment_route(current_user_id):
    data = request.json or {}
    required = ("listing_id", "vehicle_id", "start_time", "end_time")
    if not all(data.get(k) for k in required):
        return jsonify({"error": f"Missing required fields: {', '.join(required)}"}), 400
    return create_booking_payment(
        data["listing_id"], current_user_id, data["vehicle_id"],
        data["start_time"], data["end_time"],
    )


# Release a checkout hold early (renter dismissed the Stripe sheet). The sweep
# is the backstop, but releasing now frees the slot for others immediately.
#
# Authenticated: a hold_id alone used to be enough to drop anyone's slot
# mid-checkout. release_booking_hold scopes the delete to the caller.
@stripe_bp.route('/stripe/release-hold', methods=['POST'])
@token_required
def release_hold_route(current_user_id):
    data = request.json or {}
    return release_booking_hold(data.get("hold_id"), current_user_id)


# Client calls this the moment the payment sheet returns success, so the
# reservation exists immediately instead of waiting on the webhook (idempotent).
#
# Authenticated, and finalize_booking additionally checks the PaymentIntent
# belongs to the caller. Note a 401 here is a DEFERRED booking, not a lost one:
# if the session expired between paying and this call, the webhook still
# creates the reservation. Seeing these in the logs is not an emergency.
@stripe_bp.route('/stripe/finalize-booking', methods=['POST'])
@token_required
def finalize_booking_route(current_user_id):
    data = request.json or {}
    return finalize_booking(data.get("payment_intent_id"), current_user_id)


# Onboarding: create the seller's Express connected account (lazily, at payout setup).
@stripe_bp.route('/stripe/create-connect-account', methods=['POST'])
@token_required
def create_connect_account(current_user_id):
    return createConnectAccount(current_user_id)


# Onboarding: hosted Stripe onboarding link for the seller to finish KYC/bank setup.
@stripe_bp.route('/stripe/create-account-link', methods=['POST'])
@token_required
def create_account_link(current_user_id):
    return createAccountLink(current_user_id)


# Simple interstitial that immediately bounces the in-app browser back to the
# app via the deep link (nicer than a bare 302, and it renders past ngrok's free
# warning page once the user taps through).
def _deep_link_page(target: str, message: str = "Finishing up…"):
    html = f"""<!doctype html>
<html><head><meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta http-equiv="refresh" content="0;url={target}">
<title>Returning to SpotOn…</title></head>
<body style="font-family:-apple-system,BlinkMacSystemFont,sans-serif;text-align:center;padding-top:20vh;color:#1E7D46;">
<p style="font-size:18px;">{message}</p>
<a href="{target}" style="color:#1E7D46;">Tap here if you're not redirected</a>
<script>window.location.replace("{target}");</script>
</body></html>"""
    return html, 200, {"Content-Type": "text/html; charset=utf-8"}


@stripe_bp.route('/stripe/onboarding-complete', methods=['GET'])
def onboarding_return():
    # Backstop for the account.updated webhook: if the seller returned with
    # payouts enabled, release their pending reservations now even if the webhook
    # was missed. Best-effort — never block the redirect back into the app.
    user_id = request.args.get("user_id")
    if user_id:
        try:
            sync_connect_account(user_id)
        except Exception as err:  # noqa: BLE001
            print(f"[stripe] onboarding-complete sync failed for {user_id}: {err}")
    return _deep_link_page("spoton://Homescreen")


@stripe_bp.route('/stripe/onboarding-expired', methods=['GET'])
def onboarding_expired():
    return _deep_link_page("spoton://Homescreen")


# Client-invoked backstop: the app calls this after the onboarding browser closes
# so payouts sync immediately even if the webhook and the return redirect were
# both missed (idempotent with account.updated).
@stripe_bp.route('/stripe/sync-account', methods=['POST'])
@token_required
def sync_account_route(current_user_id):
    return sync_connect_account(current_user_id)


# Stripe -> SpotOn events: payment_intent.succeeded, account.updated, etc.
@stripe_bp.route('/stripe/webhook', methods=['POST'])
def stripe_webhook():
    return handle_webhook(request.get_data(), request.headers.get("Stripe-Signature", ""))


# Periodic housekeeping (invoked by pg_cron / the fallback scheduler).
# Guarded by a shared secret so it can't be triggered by arbitrary clients.
@stripe_bp.route('/stripe/run-payout-sweep', methods=['POST'])
def run_sweep():
    expected = os.getenv("SWEEP_SECRET")
    if not expected or request.headers.get("X-Sweep-Secret") != expected:
        return jsonify({"error": "unauthorized"}), 401
    return jsonify(run_payout_sweep()), 200


@stripe_bp.route('/stripe/run-notification-sweep', methods=['POST'])
def run_notifications():
    expected = os.getenv("SWEEP_SECRET")
    if not expected or request.headers.get("X-Sweep-Secret") != expected:
        return jsonify({"error": "unauthorized"}), 401
    return jsonify(run_notification_sweep()), 200

File: frontend/src/utils/stripe.ts
import { supabase } from "./supabase";
const API_IP = process.env.EXPO_PUBLIC_IP ? `https://${process.env.EXPO_PUBLIC_IP}/api` : "nulled";

export interface StripePaymentSheetParams {
    paymentIntent: string;
    customerSessionClientSecret: string;
    customer: string;
    holdId: string;
}

export interface BookingPaymentArgs {
    listing_id: string;
    renter_id: string;
    vehicle_id: string;
    start_time: string; // ISO 8601
    end_time: string;   // ISO 8601
}

/**
 * Result of attempting to start checkout.
 *  - ok:        hold acquired + PaymentIntent created; present the sheet.
 *  - unavailable: someone else is checking out / already booked this slot.
 *  - error:     anything else (network, pricing, server).
 */
export type BookingPaymentResult =
    | { status: 'ok'; params: StripePaymentSheetParams }
    | { status: 'unavailable'; message: string }
    | { status: 'error'; message: string };

const getKey = async () : Promise<string | null> => {
    const resp = await fetch(`${API_IP}/stripe/key`, {
        method: "GET",
        headers: { "ngrok-skip-browser-warning": "true" }
    });
    if (!resp.ok)
        return null;

    const data = await resp.json();
    return data.publishableKey;
}

/**
 * Hold-and-confirm checkout — call this only when the renter presses Reserve.
 * The backend acquires a short-lived slot hold and creates a platform-held
 * PaymentIntent. The reservation row itself is created by the webhook once
 * payment succeeds, so backing out of the sheet leaves nothing behind.
 * A `slot_unavailable` (409) means another renter is mid-checkout for the slot.
 */
const createBookingPayment = async (args: BookingPaymentArgs) : Promise<BookingPaymentResult> => {
    console.log(`Attempting to create booking payment @ ${API_IP}/stripe/create-booking-payment`)
    const { data: { session } } = await supabase.auth.getSession();

    let resp: Response;
    try {
        resp = await fetch(`${API_IP}/stripe/create-booking-payment`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                "ngrok-skip-browser-warning": "true",
                "Authorization": `Bearer ${session?.access_token}`
            },
            body: JSON.stringify(args)
        });
    } catch (e: any) {
        return { status: 'error', message: e?.message ?? 'Network request failed' };
    }

    if (!resp.ok) {
        const body = await resp.json().catch(() => ({} as any));
        console.log(`Error creating booking payment (${resp.status}): ${JSON.stringify(body)}`);
        if (resp.status === 409 && body?.code === 'slot_unavailable') {
            return { status: 'unavailable', message: body?.error ?? 'This spot is being booked right now.' };
        }
        return { status: 'error', message: body?.error ?? `Request failed (${resp.status})` };
    }

    const { paymentIntent, customerSessionClientSecret, customer, holdId } = await resp.json();
    console.log("Created booking payment");
    return { status: 'ok', params: { paymentIntent, customerSessionClientSecret, customer, holdId } };
}

/**
 * Create the reservation immediately after the payment sheet reports success.
 * Idempotent with the webhook (whichever runs first wins). Pass the PaymentIntent
 * client secret; the id is derived from it. Returns true if the reservation now
 * exists. Best-effort — the webhook is the backstop if this call fails.
 */
const finalizeBooking = async (paymentIntentClientSecret: string): Promise<{ ok: boolean; message?: string }> => {
    const paymentIntentId = paymentIntentClientSecret.split('_secret')[0];
    const { data: { session } } = await supabase.auth.getSession();
    try {
        const resp = await fetch(`${API_IP}/stripe/finalize-booking`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                "ngrok-skip-browser-warning": "true",
                "Authorization": `Bearer ${session?.access_token}`,
            },
            body: JSON.stringify({ payment_intent_id: paymentIntentId }),
        });
        if (!resp.ok) {
            const body = await resp.json().catch(() => ({} as any));
            const message = body?.error ?? `Request failed (${resp.status})`;
            console.log(`finalize-booking failed (${resp.status}): ${JSON.stringify(body)}`);
            return { ok: false, message };
        }
        return { ok: true };
    } catch (e: any) {
        console.log('finalize-booking network error', e);
        return { ok: false, message: e?.message ?? 'Network request failed' };
    }
}

/** Free a checkout hold immediately (renter dismissed the Stripe sheet). */
const releaseHold = async (holdId: string | null | undefined): Promise<void> => {
    if (!holdId) return;
    const { data: { session } } = await supabase.auth.getSession();
    try {
        await fetch(`${API_IP}/stripe/release-hold`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                "ngrok-skip-browser-warning": "true",
                "Authorization": `Bearer ${session?.access_token}`,
            },
            body: JSON.stringify({ hold_id: holdId }),
        });
    } catch {
        // Best-effort; the sweep will expire it anyway.
    }
}

const fetchStripeAccountId = async (user_id: string): Promise<string | null> => {
    const { data: { session } } = await supabase.auth.getSession();

    console.log(`Attempting to create stripe account for user ${user_id}...`)

    const resp = await fetch(`${API_IP}/stripe/create-connect-account`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            "ngrok-skip-browser-warning": "true",
            "Authorization": `Bearer ${session?.access_token}`
        },

        body: JSON.stringify({ user_id })
    });

    if (!resp.ok) { 
        console.log("Error fetching stripe account ID for user " + user_id);
        return null;
    }

    const { account_id } = await resp.json();
    const { error } = await supabase.from("profiles").update({ stripe_account_id: account_id}).eq("id", user_id);

    if (error) {
        console.log("Error saving stripe account ID for user " + user_id);
        return null;
    }
    console.log("Successfully saved Stripe account ID for user " + user_id)

    return account_id;
}

const fetchStripeAccountLink = async (user_id: string): Promise<string | null> => {
    const { data: { session } } = await supabase.auth.getSession();

    const resp = await fetch(`${API_IP}/stripe/create-account-link`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            "ngrok-skip-browser-warning": "true",
            "Authorization": `Bearer ${session?.access_token}`
        },
        
        body: JSON.stringify({ user_id })
    });

    if (!resp.ok) {
        console.log("Error fetching stripe account link for user " + user_id);
        return null;
    }

    const { account_link_url } = await resp.json();
    return account_link_url;
}

//checks if user has an existing stripe connect account.
const userHasStripeAccount = async (userId: string): Promise<boolean> => {
  const { data, error } = await supabase.from('profiles').select('stripe_account_id').eq('id', userId).single();

  if (error) {
    console.error('Error checking for Stripe account:', error.message);
    return false;
  }

  return !!data?.stripe_account_id;
};

export const stripe = { getKey, createBookingPayment, finalizeBooking, releaseHold, fetchStripeAccountId, fetchStripeAccountLink, userHasStripeAccount }
